android/openthread_commissioner/service/src/main/java/io/openthread/commissioner/service/BorderAgentDiscoverer.java

```diff
@@ -0,0 +1,238 @@
+
+    void onBorderAgentLost(String discriminator);
+  }
+
+  @RequiresPermission(permission.INTERNET)
+  public BorderAgentDiscoverer(Context context, BorderAgentListener borderAgentListener) {
+    WifiManager wifi = (WifiManager) context.getSystemService(Context.WIFI_SERVICE);
+    wifiMulticastLock = wifi.createMulticastLock("multicastLock");
+
+    nsdManager = (NsdManager) context.getSystemService(Context.NSD_SERVICE);
+
+    this.borderAgentListener = borderAgentListener;
+  }
+
+  public void start() {
+    if (isScanning) {
+      Log.w(TAG, "the Border Agent discoverer is already running!");
+      return;
+    }
+
+    isScanning = true;
+
+    wifiMulticastLock.setReferenceCounted(true);
+    wifiMulticastLock.acquire();
+
+    startResolver();
+    nsdManager.discoverServices(
+        BorderAgentDiscoverer.SERVICE_TYPE, NsdManager.PROTOCOL_DNS_SD, this);
+  }
+
+  private void startResolver() {
+    NsdManager.ResolveListener listener =
+        new NsdManager.ResolveListener() {
+          @Override
+          public void onResolveFailed(NsdServiceInfo serviceInfo, int errorCode) {
+            Log.e(
+                TAG,
+                String.format(
+                    "failed to resolve service %s, error: %d, this=%s",
+                    serviceInfo.toString(), errorCode, this));
+            isResolvingService.set(false);
+          }
+
+          @Override
+          public void onServiceResolved(NsdServiceInfo serviceInfo) {
+            BorderAgentInfo borderAgent = getBorderAgentInfo(serviceInfo);
+            if (borderAgent != null) {
+              Log.d(TAG, "successfully resolved service: " + serviceInfo.toString());
+              Log.d(
+                  TAG,
+                  "successfully resolved service: " + serviceInfo.getHost().getCanonicalHostName());
+              borderAgentListener.onBorderAgentFound(borderAgent);
+            }
+            isResolvingService.set(false);
+          }
+        };
+
+    Log.d(TAG, "mDNS resolve listener is " + listener);
+
+    if (executor.isTerminated()) {
+      executor = Executors.newSingleThreadExecutor();
+    }
+
+    executor.submit(
+        () -> {
+          while (true) {
+            if (!isResolvingService.get()) {
+              NsdServiceInfo serviceInfo = unresolvedServices.take();
+
+              isResolvingService.set(true);
+              nsdManager.resolveService(serviceInfo, listener);
+            }
+          }
+        });
+  }
+
+  private void stopResolver() {
+    if (!executor.isTerminated()) {
+      executor.shutdownNow();
+    }
+    isResolvingService.set(false);
+    unresolvedServices.clear();
+  }
+
+  public void stop() {
+    if (!isScanning) {
+      Log.w(TAG, "the Border Agent discoverer has already been stopped!");
+      return;
+    }
+
+    nsdManager.stopServiceDiscovery(this);
+    stopResolver();
+
+    if (wifiMulticastLock != null) {
+      wifiMulticastLock.release();
```

Comment:
```suggestion
      wifiMulticastLock.release();
      wifiMulticastLock = null;
```

Reply:
`release` is to release the lock but not the object itself. we may re-acquire this lock later.